upa/contracts/Deposits.sol

```diff
@@ -0,0 +1,283 @@
+    }
+
+    // Data to be signed by the aggregator
+    struct AggregationAgreement {
+        bytes32 submissionId;
+        uint256 expirationBlockNumber;
+        uint256 fee;
+        address submitterAddress;
+    }
+
+    /// Aggregtor fee claims are sent here
+    address public immutable AGGREGATOR_ADDRESS;
+
+    /// The UPA verifier contract.
+    IUpaVerifier public immutable UPA_VERIFIER;
+
+    /// Submitter accounts
+    mapping(address => SubmitterAccount) public accounts;
+
+    /// Mapping from an `AggregationAgreement` hash to whether a refund claim
+    /// has been issued for that agreement.
+    mapping(bytes32 => bool) public submissionRefunded;
+
+    // Withdrawal notice is about 2 days
+    uint256 public constant WITHDRAWAL_NOTICE_BLOCKS = 15000;
+
+    bytes32 private constant SIGNED_REQUEST_DATA_TYPE_HASH =
+        keccak256(
+            // solhint-disable-next-line
+            "SignedRequestData(bytes32 submissionId,uint256 expirationBlockNumber,uint256 totalFee)"
+        );
+
+    bytes32 private constant SIGNED_AGGREGATION_AGREEMENT_TYPE_HASH =
+        keccak256(
+            // solhint-disable-next-line
+            "AggregationAgreement(bytes32 submissionId,uint256 expirationBlockNumber,uint256 fee,address submitterAddress)"
+        );
+
+    /// Gas per Ethereum transaction.
+    uint256 private constant GAS_PER_TRANSACTION = 21000;
+
+    /// Gas for the SSTORE used to refund fees.
+    uint256 private constant GAS_PER_REFUND_SSTORE = 5000;
+
+    // TODO: collateral, should be some amount larger than user deposits.
+
+    constructor(
+        string memory name,
+        string memory version,
+        address _aggregatorAddress,
+        address _upaVerifierAddress
+    ) EIP712(name, version) {
+        require(_aggregatorAddress != address(0), AggregatorAddressIsZero());
+        require(_upaVerifierAddress != address(0), UpaAddressIsZero());
+        AGGREGATOR_ADDRESS = _aggregatorAddress;
+        UPA_VERIFIER = IUpaVerifier(_upaVerifierAddress);
+    }
+
+    /// Return the total wei deposited by this submitter, including fees
+    /// that have been claimed by the aggregator, and a flag indicating if
+    /// there is a pending withdrawal initiated.
+    function getSubmitterTotal(
+        address submitter
+    ) external view returns (uint256 total, uint256 withdrawalInitBlock) {
+        SubmitterAccount storage account = accounts[submitter];
+        total = account.balance + account.totalClaimed;
+        withdrawalInitBlock = account.pendingWithdrawalInitializedAtBlock;
+    }
+
+    function balance(address account) external view returns (uint256) {
+        return accounts[account].balance;
+    }
+
+    function pendingWithdrawalInitializedAtBlock(
+        address account
+    ) external view returns (uint256) {
+        return accounts[account].pendingWithdrawalInitializedAtBlock;
+    }
+
+    /// Top-up submitter's ETH balance
+    function deposit() external payable {
```

Comment:
@dysfunctor Should we just have a fallback function, so people can top up with just a send tx?

Reply:
Collateral deposits can use the same approach (so fallback has to check whether msg.sender == owner), or a separate function.

Reply:
#1104